subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultGroovySourceSet.java

```diff
@@ -18,31 +18,36 @@
 import groovy.lang.Closure;
 import org.gradle.api.Action;
 import org.gradle.api.file.SourceDirectorySet;
+import org.gradle.api.internal.file.DefaultSourceDirectorySet;
+import org.gradle.api.internal.file.FileCollectionFactory;
+import org.gradle.api.internal.file.collections.DirectoryFileTreeFactory;
 import org.gradle.api.model.ObjectFactory;
 import org.gradle.api.reflect.HasPublicType;
 import org.gradle.api.reflect.TypeOf;
 import org.gradle.api.tasks.GroovySourceSet;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.Factory;
 
 import javax.annotation.Nullable;
 
 import static org.gradle.api.reflect.TypeOf.typeOf;
 import static org.gradle.util.internal.ConfigureUtil.configure;
 
-public class DefaultGroovySourceSet implements GroovySourceSet, HasPublicType {
-    private final SourceDirectorySet groovy;
+public class DefaultGroovySourceSet extends DefaultSourceDirectorySet implements GroovySourceSet, HasPublicType {
```

Comment:
OK, I think I understand why you're extending `DefaultSourceDirectorySet`.

If someone is doing this in Groovy DSL:
```
sourceSets {
    main {
        groovy {
            srcDir = ...
        }
    }
}
```

Replacing the convention `GroovySourceSet` with the extension version breaks this because the `groovy` call here is doing something different. It's configuring the `GroovySourceSet` extension and _not_ calling `groovy(Action<? super SourceDirectorySet> configureAction)`. If you didn't extend `DefaultSourceDirectorySet`, you'd have to at least implement `SourceDirectorySet` to keep things working.

I think this has the side effect of making this valid now (in Groovy):
```
sourceSets {
    main {
        groovy {
            groovy {
                srcDir = ...
            }
        }
    }
}
```

This is valid, but weird. If we wanted to deprecate the `groovy(Action)` methods to discourage this, that causes a new problem.

In Kotlin, people do something like this:
```
sourceSets {
    main {
        withConvention<GroovySourceSet> {
            groovy {
                srcDir = ...
            }
        }
    }
}
```

And with this change, that could be replaced with:
```
sourceSets {
    main {
        configure<GroovySourceSet> {
            groovy {
                srcDir = ...
            }
        }
    }
}
```

And since we're registering the `GroovySourceSet` as an extension, we get an accessor for that (now we get exactly the same as Groovy in the "weird" case above):
```
sourceSets {
    main {
        groovy {
            groovy {
                srcDir = ...
            }
        }
    }
}
```

Unfortunately, since `GroovySourceSet` doesn't extend `SourceDirectorySet`, you can't do this:
```
sourceSets {
    main {
        groovy {
            srcDir = ...
        }
    }
}
```
...which I think would be our preferred syntax in both DSLs and if we deprecate/remove the `groovy()` methods, you won't be able to configure this in Kotlin at all.

I think you're on the right track with extending `DefaultSourceDirectorySet`, but I think we might have to do something a little more.

The issue with extending `DefaultSourceDirectorySet` for the Groovy convention is that it also adds properties from `SourceDirectorySet` unintentionally due to the way we delegate things to conventions.

e.g., this starts working:
```
sourceSets {
    main {
        println srcDirs
    }
}
```
because we find the `srcDirs` property on the `GroovySourceSet` convention object.


I think our goal should be this in both DSLs:
```
sourceSets {
    main {
        groovy {
            srcDir = ...
        }
    }
}
```

This is what I was thinking:
1. Add an extension called `groovy` of type `DefaultSourceDirectorySet` and with a public type of `SourceDirectorySet` to the source set like you're doing in this change.
2. Keep `DefaultGroovySourceSet` as a convention, but change it around so that it returns the extension from (1) as the "groovy" source set field and do not extend `DefaultSourceDirectorySet`.
3. Use the preferred syntax in all samples/docs/tests for both DSLs
4. Deprecate `GroovySourceSet`. We can add nagging when we deprecate `getConventions`
5. Remove the use of `GroovySourceSet.allGroovy` (I think "allGroovy" here really means "only Groovy source files")

WDYT? Does this break something else?

Reply:
I've thought of something else that complicates my suggestion:
https://gradle.slack.com/archives/CDYR952SH/p1620863077258700

I'm leaning towards introducing a new type (like `GroovySourceDirectorySet`) for each extension to separate the convention and the extension completely.

Reply:
I think this has the side effect of making this valid now (in Groovy):
```
sourceSets {
    main {
        groovy {
            groovy {
                srcDir = ...
            }
        }
    }
}
```

Reply:
At the same time, the Kotlin DSL is weird, I didn't realize that:
<img width="616" alt="image" src="https://user-images.githubusercontent.com/419883/118093643-06951300-b3ce-11eb-8874-b8a3f51b3a47.png">

Reply:
OK, I've thoroughly read your comments. It all makes sense; I'll give it a try.

Reply:
I did an initial round on this and it looks good. The only weird thing is that nested Groovy statements are valid in existing Gradle releases with both Groovy and Kotlin DLSs. 
```
sourceSets {
    main {
        withConvention(GroovySourceSet::class) {
            groovy {
                groovy {
                    setSrcDirs(listOf("custom-groovy"))
                }
            }
        }
    }
}
```

Reply:
In case there's anyone else following this, we decided to add specific types for each source set type.